### PR TITLE
[layouts][server] Composition -> layout

### PR DESCRIPTION
**Pull request by elpaso**

### CODE REVIEW

File: src/server/services/wms/qgswmsgetcapabilities.cpp [hunk 1]
```diff
@@ -624,33 +624,34 @@ namespace QgsWms
...
         continue;
 
       QDomElement composerTemplateElem = doc.createElement( QStringLiteral( "ComposerTemplate" ) );
-      composerTemplateElem.setAttribute( QStringLiteral( "name" ), composer->name() );
+      composerTemplateElem.setAttribute( QStringLiteral( "name" ), layout->name() );
 
       //get paper width and hight in mm from composition
-      composerTemplateElem.setAttribute( QStringLiteral( "width" ), composer->paperWidth() );
-      composerTemplateElem.setAttribute( QStringLiteral( "height" ), composer->paperHeight() );
+      composerTemplateElem.setAttribute( QStringLiteral( "width" ), layout->width() );
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsgetcapabilities.cpp]: You'll need to take the first page size here - the width and height functions are unfortunately inherited from QGraphicsScene and give the whole scene bounds dimensions. This includes all items, including cosmetic things like guides, page shadow, etc.

File: src/server/services/wms/qgswmsgetcapabilities.cpp [hunk 2]
```diff
@@ -624,33 +624,34 @@ namespace QgsWms
...
+      composerTemplateElem.setAttribute( QStringLiteral( "height" ), layout->height() );
 
       //add available composer maps and their size in mm
-      QList<const QgsComposerMap *> composerMapList = composer->composerMapItems();
-      QList<const QgsComposerMap *>::const_iterator cmIt = composerMapList.constBegin();
+      QList<QgsLayoutItemMap *> composerMapList;
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsgetcapabilities.cpp]: Can we s/composer/layout/ here?

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 3]
```diff
@@ -354,284 +360,266 @@ namespace QgsWms
...
-          //remove map from composition if not referenced by the request
-          c->removeItem( map );
-          delete map;
-          continue;
-        }
+      //map extent is mandatory
+      if ( !cMapParams.mHasExtent )
+      {
+        //remove map from composition if not referenced by the request
+        c->removeItem( map );
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsrenderer.cpp]: removeLayoutItem

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 4]
```diff
@@ -354,284 +360,266 @@ namespace QgsWms
...
+        delete map;
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsrenderer.cpp]: Not required for removeLayoutItem

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 5]
```diff
@@ -354,284 +360,266 @@ namespace QgsWms
...
-          //remove exported labels referenced in the request
-          //but with empty string
-          c->removeItem( label );
-          delete label;
-          continue;
-        }
+    // Labels
+    QList<QgsLayoutItemLabel *> labels;
+    c->layoutItems<QgsLayoutItemLabel>( labels );
+    for ( auto &label : labels )
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsrenderer.cpp]: qgis::as_const( labels )

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 6]
```diff
@@ -354,284 +360,266 @@ namespace QgsWms
...
-    // Update composer items
-    QList<QgsComposerItem *>::iterator itemIt = itemList.begin();
-    for ( ; itemIt != itemList.end(); ++itemIt )
+    // Maps are configured first
+    QList<QgsLayoutItemMap *> maps;
+    c->layoutItems<QgsLayoutItemMap>( maps );
+    // Layout maps now use a string UUID as "id", let's assume that the first map
+    // has id 0 and so on ...
+    int mapId = 0;
+    for ( auto &map : maps )
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsrenderer.cpp]: qgis::as_const(maps)

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 7]
```diff
@@ -354,284 +360,266 @@ namespace QgsWms
...
-        }
+      label->setText( labelParam );
+    }
 
-        QUrl newUrl( url );
-        html->setUrl( newUrl );
+    // HTMLs
+    QList<QgsLayoutItemHtml *> htmls;
+    c->layoutObjects<QgsLayoutItemHtml>( htmls );
+    for ( auto &html : htmls )
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsrenderer.cpp]: qgis::as_const( htmls )

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 8]
```diff
@@ -354,284 +360,266 @@ namespace QgsWms
...
+      QUrl newUrl( url );
+      html->setUrl( newUrl );
+      html->update();
+    }
+
+
+    // legends
+    QList<QgsLayoutItemLegend *> legends;
+    c->layoutItems<QgsLayoutItemLegend>( legends );
+    for ( auto &legend : legends )
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsrenderer.cpp]: as_const

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 9]
```diff
@@ -354,284 +360,266 @@ namespace QgsWms
...
 
-        currentLegend->setAutoUpdateModel( false );
+        legend->setAutoUpdateModel( false );
 
         // get model and layer tree root of the legend
-        QgsLegendModel *model = currentLegend->model();
+        QgsLegendModel *model = legend->model();
         QStringList layerSet;
-        Q_FOREACH ( QgsMapLayer *layer, map->layers() )
+        for ( const auto &layer : map->layers() )
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsrenderer.cpp]: Need to make a const copy of layers here to iterate over

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 10]
```diff
@@ -640,9 +628,9 @@ namespace QgsWms
         // get layerIds find in the layer tree root
         QStringList layerIds = root->findLayerIds();
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsrenderer.cpp]: const QStringList

File: tests/src/python/qgis_wrapped_server.py [hunk 11]
```diff
@@ -10,6 +10,11 @@
 
   ?MAP=/path/to/projects.qgs&SERVICE=XYZ&X=1&Y=0&Z=1&LAYERS=world
 
+Note that multi threading in QGIS server is not officially supported and
+it is not supposed to work in any case
+
+Set MULTITHREADING environment varialbe to 1 to activate.
```
Comment (nyalldawson) [inline tests/src/python/qgis_wrapped_server.py]: Variable

**Review by nyalldawson — APPROVED** [review]
Looks good, +1 to merge when ready

**Comment by tudorbarascu** [comment]
@elpaso Thanks again for the server work! Is this supposed to touch also the exclude composers from the server part? Tested and it's not working.

**Comment by elpaso** [comment]
Is there any test for that?
I don't see why it should be broken but if you say it is, I trust you :)

**Comment by elpaso** [comment]
@tudorbarascu btw I tested the exclude and it works for me (the excluded template is not in the GetProjectsettings response)

**Comment by tudorbarascu** [comment]
@elpaso Thanks for looking. I just rebuild, created a new project from scratch, added some layouts.. but I cannot select them from a combobox like in version 2. (Haven't tested if it works by filling the template name manually).
bad
Thanks a lot!

**Comment by Gustry** [comment]
Shouldn't rename all "composers" to "layouts" in your screenshot for consistency?

**Comment by elpaso** [comment]
@Gustry not sure, because "layout" may just be the internal code name, "composer" might be more understandable and I don't see any advantage in renaming it from the user perspective: @nyalldawson ?

I also did not rename the XML tag in getprojectsettings, that would make it not compatible with all webgis software out there.

@tudorbarascu I tested manually enter and that works. I'll see what misses for the combo.

**Comment by Gustry** [comment]
screen shot 2018-01-11 at 16 04 31

@elpaso AFAIK, "Layout" is already replaced in a lot places (window title, layout manager, layout properties...)

**Comment by elpaso** [comment]
well, in that case I guess it needs renamed ... btw this PR was focused on the QGIS server side of thing not on the GUI, but I'll take care of it.

**Comment by nyalldawson** [comment]
@elpaso

> not sure, because "layout" may just be the internal code name, "composer" might be more understandable and I don't see any advantage in renaming it from the user perspective

When this work was all being scoped (2 years ago?) there was a discussion on the lists about renaming compositions to layouts within the 2.x cycle. There was lots of support for this, but it was decided to delay the UI change until the new engine and 3.0.

The main reason is that "composition"/"composer" is quite an awkward term in English, usually associated only with music.

**Pull request merged**